### ISSUE TITLE
get-update-detail should return entries from NEWS in apt backend

### ISSUE THREAD

**Issue by sidt4** [PackageKit/PackageKit#654]
It appears that `get-update-detail` returns entries from the NEWS file in `dnf` backend.

But, in `apt` backend we return the Debian downstream `ChangeLog`. I am not sure if this Debian `ChangeLog` is actually being used by any PK clients. I would suggest to replace the Debian `ChangeLog` entries with the entries from NEWS file, which are more meaningful to PK clients. If not replace,  at least an option should be provided to select between `NEWS` and `Debian ChangeLog`

**Comment by ximion** [comment]
The changelog data is used by Apper, KDE Discover and GNOME PackageKit.
What do you mean with NEWS file? The `NEWS.Debian` file? Most packages don't even have that, it's fairly rare (and usually exists to alert technical users of breaking changes, so it's very verbose).

**Comment by sidt4** [comment]
> What do you mean with NEWS file? The NEWS.Debian file? Most packages don't even have that, it's fairly rare (and usually exists to alert technical users of breaking changes, so it's very verbose).

No. I meant the `NEWS` / `ChangeLog` file which is where the upstream maintainers update the release notes for their packages. Refer below for more details:

```
$pkcon get-update-detail audit
Resolving                     [=========================]         More than one package matches:
1. audit-3.1.2-1.fc38.x86_64 [updates]
2. audit-3.1-2.fc38.x86_64 [fedora]
Please choose the correct package: 1
Getting update details        [=========================]         
Starting                      [=========================]         
Finished                      [=========================]         
Details about the update:
 Package: audit-3.1.2-1.fc38.x86_64
 Update text: Various bugfixes, updated lookup tables for the 6.5 kernel, and added some new python functions.
 Changes: 
 State: stable
 Issued: 
 Updated: 
```

Where does the following text come from ?

```
 Update text: Various bugfixes, updated lookup tables for the 6.5 kernel, and added some new python functions.
```

I checked the `Update text:` for a package a while ago, and it matched the `NEWS` file shipped with the package. So, I thought it was from the `NEWS` file ( though was wondering how it's possible to access what's inside a compressed source archive ). But the `NEWS` file for `audit` is empty. And, below are the contents from the `ChangeLog` file shipped with the package.

```
3.1.2
- When processing a run level change, make auditd exit
- In auditd, fix return code when rules added in immutable mode
- In auparse, when files are given, also consider EUID for access
- Auparse now interprets unnamed/anonymous sockets (Enzo Matsumiya)
- Disable Python bindings from setting rules due to swig bug (S. Trofimovich)
- Update all lookup tables for the 6.5 kernel
- Don't be as paranoid about auditctl -R file permissions
- In ausearch, correct subject/object search to be an and if both are given
- Adjust formats for 64 bit time_t
- Fix segfault in python bindings around the feed API
- Add feed_has_data, get_record_num, and get/goto_field_num to python bindings
```

So, it appears that there is manual update of the `Update text:` for each package in fedora ?

**Comment by sidt4** [comment]
> The changelog data is used by Apper, KDE Discover and GNOME PackageKit.

No. I didn't mean whether the `get-update-detail` response is being used in any PK client.

I meant whether the  `get-update-detail` response ( a.ka, `Debian downstream ChangeLog` ), which contains real technical information about the changes made by Debian maintainers to the package ( which most of the times have nothing to do with the actual upstream package ), is actually serving any use case in any of the PK clients.

**Comment by ximion** [comment]
Ah! Debian doesn't have that concept, so since the metadata simply doesn't exist in Debian (and doesn't exist in a standardized format upstream anyway, unless the project support AppStream in which case tools can get the info from AppStream instead) we can't implement this feature at all.

It's a Fedora-ism, as far as I know, that not all backends can support.

> which contains real technical information about the changes made by Debian maintainers to the package

That's not entirely true - many Debian maintainers put highlights from the upstream release and especially noteworthy changes in there too (e.g. https://metadata.ftp-master.debian.org/changelogs//main/g/gnome-software/gnome-software_45~beta-1_changelog has some highlighted bug fixes, but there's better examples too).

**Comment by sidt4** [comment]
Please don't get me wrong. I do find Debian `ChangeLog` useful as a developer ( 80% ) and user ( 20% ). The use case here is to present what's changed between version `x` to `y` in package `Foo` to the user in GNOME Software UI. And we cannot display the following to a normal desktop user.

```
gnome-software (44.3-2) unstable; urgency=medium

  * Remove an obsolete maintscript entry
  * Release to unstable

 -- Jeremy Bícha <jbicha@ubuntu.com>  Sun, 23 Jul 2023 16:22:56 -0400

gnome-software (44.3-1) experimental; urgency=medium

  * Team upload
  * New upstream release
  
 -- Jarred Wilson <jarred.wilson@canonical.com>  Tue, 11 Jul 2023 14:40:17 -0400
  ```

Users expect something like the following in release notes in GNOME Software:

```
What's new in version 44.3:

* Software updates are now faster
* Fixed an issue where Software update fails sometimes in Debian / Ubuntu
* Fixed an issue which caused app icons to be not displayed in top banner 
* .....
``` 

Desktop apps have appdata / metainfo files. Non-desktop apps update `NEWS` / `ChangeLog` files, which have all the information needed by GNOME Software, but they're not present in a standard global database / file. Rather the `NEWS` / `ChangeLog` resides within the source `tar.gz`. Maybe, it's possible to read the `NEWS` / `ChangeLog` files, once they're installed on the system, but for remote packages ( for update release notes ), this is not possible.

I think we should do an https://appstream.debian.org equivalent for non-desktop packages too, or  make non-desktop packages to use appdata / metainfo files.

**Comment by ximion** [comment]
> I think we should do an https://appstream.debian.org equivalent for non-desktop packages too, or make non-desktop packages to use appdata / metainfo files.

https://www.freedesktop.org/software/appstream/docs/sect-Metadata-ConsoleApplication.html
https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#sect-Metadata-GenericComponent

People just need to add the metadata. The NEWS files don't have any defined format (GNU doesn't even make any strong recommendation), so they can't even be parsed. They are also not always installed on Debian systems, and definitely not extracted as metadata that we could fetch.

So, I'm afraid this feature request can't be implemented for the APT backend, and instead people will simply need to add more AppStream metadata.

Also, tbh, I don't think people will look at changes in individual library packages that are installed. And people who are interested in that will be perfectly fine with Debian changelogs.

**Comment by sidt4** [comment]
> They are also not always installed on Debian systems,

```
$ ls  /usr/share/doc/*/NEWS.gz | wc -w
746
```

```
$ ls  /usr/share/doc/*/NEWS.gz
...
...
/usr/share/doc/gnome-system-log/NEWS.gz               /usr/share/doc/libjson-glib-1.0-0/NEWS.gz             /usr/share/doc/seahorse/NEWS.gz
/usr/share/doc/gnome-system-monitor/NEWS.gz           /usr/share/doc/libkpathsea6/NEWS.gz                   /usr/share/doc/sed/NEWS.gz
/usr/share/doc/gnome-taquin/NEWS.gz                   /usr/share/doc/libksba8/NEWS.gz                       /usr/share/doc/shared-mime-info/NEWS.gz
/usr/share/doc/gnome-tetravex/NEWS.gz                 /usr/share/doc/liblrdf0/NEWS.gz                       /usr/share/doc/sharutils/NEWS.gz
/usr/share/doc/gnome-themes-extra/NEWS.gz             /usr/share/doc/liblsan0/NEWS.gz                       /usr/share/doc/shotwell-dbgsym/NEWS.gz
/usr/share/doc/gnome-tweaks/NEWS.gz                   /usr/share/doc/libltdl7-dbgsym/NEWS.gz                /usr/share/doc/shotwell/NEWS.gz
/usr/share/doc/gnome-usage/NEWS.gz                    /usr/share/doc/libltdl7/NEWS.gz                       /usr/share/doc/simple-scan-dbgsym/NEWS.gz
/usr/share/doc/gnome-user-docs/NEWS.gz                /usr/share/doc/liblz1/NEWS.gz                         /usr/share/doc/simple-scan/NEWS.gz
/usr/share/doc/gnome-user-share/NEWS.gz               /usr/share/doc/liblzma5-dbgsym/NEWS.gz                /usr/share/doc/smartmontools-dbgsym/NEWS.gz
/usr/share/doc/gnome-weather/NEWS.gz                  /usr/share/doc/liblzma5/NEWS.gz                       /usr/share/doc/smartmontools/NEWS.gz
...
...

```
> Also, tbh, I don't think people will look at changes in individual library packages that are installed. And people who are interested in that will be perfectly fine with Debian changelogs.

Agree. One of the main motivations here is that many desktop applications still update `NEWS` / `ChangeLog` file and not Appdata / metainfo files ( some don't even have a Appdata / metainfo file ). So, I was wondering if we can fix the gap here with the data available to us, until most apps have moved to using Appdata. And if we can read `NEWS` / `ChangeLog` files for desktop apps, that can be used for console applications /  libraries as well ( without any additional effort ).

**Comment by ximion** [comment]
> So, I was wondering if we can fix the gap here with the data available to us

We can't, really. This is the kind of metadata available to a Debian system: https://metadata.ftp-master.debian.org/changelogs//main/s/systemd/ - the NEWS files there are actually a Debian format to alert the user of specific changes, and are not what you want.
Upstream NEWS files that are installed, as you listed, are picked up as *documentation*, and are installed as such (but the package maintainer can not do that, it's entirely their choice). We do not have access to them prior to installation.

People should just write MetaInfo files. `appstreamcli` can help with that, ages ago I blogged about it even: https://blog.tenstral.net/2020/03/maintain-release-info-easily-in-metainfo-files.html (a bit outdated now, but still mostly correct blogpost).
So, if people run `news-to-metainfo` as part of their buildprocess, they don't even need to touch XML.

Or, if you have some time and Perl knowledge, you could implement the reverse of https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=954834 to have Debian's tooling do the job - in that case you may be in for a bad time though, as NEWS files very wildly and have no defined format.

**Comment by sidt4** [comment]
> We can't, really. This is the kind of metadata available to a Debian system: https://metadata.ftp-master.debian.org/changelogs//main/s/systemd/ - the NEWS files there are actually a Debian format to alert the user of specific changes, and are not what you want.

Yeah .. they're downstream NEWS files which we're not interested in.

>  Upstream NEWS files that are installed, as you listed, are picked up as documentation, and are installed as such (but the package maintainer can not do that, it's entirely their choice). We do not have access to them prior to installation.

Right. I think we can use `appstreamcli news-to-metainfo $UPSTREAM_NEWS_FILE_AFTER_INSTALLATION` at runtime by apps like GNOME Software when there is no MetaInfo for `<pkg>`, before declaring a "No description available".

> People should just write MetaInfo files. appstreamcli can help with that, ages ago I blogged about it even: https://blog.tenstral.net/2020/03/maintain-release-info-easily-in-metainfo-files.html (a bit outdated now, but still mostly correct blogpost). So, if people run news-to-metainfo as part of their buildprocess, they don't even need to touch XML.

The link looks informative.

Honestly I didn't know how useful Appstream metainfo files can be until I started using GNOME Software. So, yes there is less awareness on this. The push to move to Appdata has to be constantly applied and for that we need good accurate documentation with pictures available to point to users who actually are interested in switching, whenever that is. The documentation should demonstrate how tags in metainfo files are actually rendered in an app like GNOME Software. And for that we need better than just a blog / link ( Maybe,  http://use-appstream-in-my.app ?  )
 
> Or, if you have some time and Perl knowledge, you could implement the reverse of https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=954834 to have Debian's tooling do the job - in that case you may be in for a bad time though, as NEWS files very wildly and have no defined format.

I'll look into this when I get some free time.

Thanks for the pointers!

**Comment by sidt4** [comment]
> The documentation should demonstrate how tags in metainfo files are actually rendered in an app like GNOME Software. 

The below talk by Cassidy James in GUADEC 2023.

https://www.youtube.com/watch?v=jIDk0frev7M&t=719s
